File: elixir/lib/2023/07_camel_cards/aoc202307.ex
defmodule AOC2023.Day07 do
  @moduledoc """
  Advent of Code 2023, day 7: Camel Cards.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.map(&String.split/1)
    |> Enum.map(fn [cards, bid] -> {String.to_charlist(cards), String.to_integer(bid)} end)
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    data
    |> Enum.map(fn {cards, bid} ->
      {rank_hand(cards, &find_hand_type/1, ~c"23456789TJQKA"), bid}
    end)
    |> Enum.sort()
    |> Enum.with_index(1)
    |> Enum.map(fn {{_, bid}, rank} -> bid * rank end)
    |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    data
    |> Enum.map(fn {cards, bid} ->
      {rank_hand(cards, &find_hand_type_w_joker/1, ~c"J23456789TQKA"), bid}
    end)
    |> Enum.sort()
    |> Enum.with_index(1)
    |> Enum.map(fn {{_, bid}, rank} -> bid * rank end)
    |> Enum.sum()
  end

  @doc """
  Rank a poker hand.

  First, rank it based on hand type (five of a kind, four of a kind, etc.), then
  rank individual cards.

  ## Examples:

      iex> rank_hand(~c"32T3K")
      [2, 1, 0, 8, 1, 11]
      iex> rank_hand(~c"T55J5")
      [4, 8, 3, 3, 9, 3]
  """
  def rank_hand(cards, hand_type \\ &find_hand_type/1, order \\ ~c"23456789TJQKA") do
    [hand_type.(cards) | cards |> Enum.map(fn card -> Enum.find_index(order, &(&1 == card)) end)]
  end

  @doc """
  Find the hand type of the cards.

  Returns a number that can be used to rank based on hand type.

  ## Examples:

      iex> find_hand_type(~c"32T3K")
      2
      iex> find_hand_type(~c"T55J5")
      4
  """
  def find_hand_type(cards) do
    num_cards = cards |> Enum.frequencies() |> Map.values() |> Enum.sort()

    case num_cards do
      [1, 1, 1, 1, 1] -> 1
      [1, 1, 1, 2] -> 2
      [1, 2, 2] -> 3
      [1, 1, 3] -> 4
      [2, 3] -> 5
      [1, 4] -> 6
      [5] -> 7
    end
  end

  @doc """
  Find the hand type of the cards, taking jokers into account.

  Returns a number that can be used to rank based on hand type.

  ## Examples:

      iex> find_hand_type_w_joker(~c"32T3K")
      2
      iex> find_hand_type_w_joker(~c"T55J5")
      6
      iex> find_hand_type_w_joker(~c"JJJJJ")
      7
  """
  def find_hand_type_w_joker(cards) do
    {best_card, _} =
      cards |> Enum.frequencies() |> Map.put(?J, 0) |> Enum.max_by(fn {_key, value} -> value end)

    find_hand_type(
      Enum.map(cards, fn
        ?J -> best_card
        card -> card
      end)
    )
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2023/07_camel_cards/aoc202307.py
"""AoC 7, 2023: Camel Cards."""

# Standard library imports
import collections
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    data = []
    for line in puzzle_input.split("\n"):
        cards, bid = line.split()
        data.append((cards, int(bid)))
    return data


def part1(data):
    """Solve part 1."""
    cards = sorted(((rank_cards(cards), bid) for cards, bid in data), reverse=True)
    return sum(rank * bid for rank, (_, bid) in enumerate(cards, start=1))


def part2(data):
    """Solve part 2."""
    cards = sorted(
        ((rank_cards(cards, use_joker=True), bid) for cards, bid in data), reverse=True
    )
    return sum(rank * bid for rank, (_, bid) in enumerate(cards, start=1))


def rank_cards(cards, use_joker=False):
    """Create a tuple that can be used to sort cards.

    ## Examples:

    >>> rank_cards("2345J", use_joker=False)
    (7, 12, 11, 10, 9, 3)
    >>> rank_cards("2345J", use_joker=True)
    (6, 11, 10, 9, 8, 12)
    >>> rank_cards("KJKKJ", use_joker=False)
    (3, 1, 3, 1, 1, 3)
    >>> rank_cards("KJKKJ", use_joker=True)
    (1, 1, 12, 1, 1, 12)
    """
    order = "AKQT98765432J" if use_joker else "AKQJT98765432"
    return (classify(cards, use_joker=use_joker),) + tuple(
        order.index(card) for card in cards
    )


def classify(cards, use_joker):
    """Classify poker hand.

    Returns a rank for each kind of hand.

    ## Examples:

    >>> classify("2345J", use_joker=False)
    7
    >>> classify("2345J", use_joker=True)
    6
    >>> classify("KJKKJ", use_joker=False)
    3
    >>> classify("KJKKJ", use_joker=True)
    1
    """
    if use_joker:
        card_counts = collections.Counter(cards)
        if "J" in card_counts:
            common, _ = collections.Counter(card_counts | {"J": 0}).most_common(1)[0]
            return classify(cards.replace("J", common), use_joker=False)

    counts = sorted(collections.Counter(cards).values())
    match counts:
        case [5]:  # Five of a kind
            return 1
        case [1, 4]:  # Four of a kind
            return 2
        case [2, 3]:  # Full house
            return 3
        case [1, 1, 3]:  # Three of a kind
            return 4
        case [1, 2, 2]:  # Two pairs
            return 5
        case [1, 1, 1, 2]:  # One pair
            return 6
        case [1, 1, 1, 1, 1]:  # High card
            return 7
        case _:
            raise ValueError(f"couldn't classify {cards}")


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
